### ISSUE TITLE
run_format_code(check_only=True) reports success when isort silently skips files

### ISSUE THREAD

**Issue by MarcusJellinghaus** [MarcusJellinghaus/mcp-code-checker#221]
## Summary

`run_format_code(check_only=True)` reports a clean result when isort has silently failed to check part of the repository. isort emits a warning per file it gives up on and **exits 0**, so the tool presents a compromised run as a passing one.

The fault is confined to **check mode** — apply mode does not exhibit it (see *Mode asymmetry* below). Note that the fix's parsing is deliberately **not** mode-gated: an apply-mode run that ever did warn would set `success = False` and halt before black via `runner.py:63`. That is intended behaviour, not a regression — see the Fail-fast row in Decisions.

The underlying fault is an isort bug, not this server's. What is in scope here is that `run_format_code` gives no indication anything went wrong.

## Observed behaviour

`run_format_code(steps=["isort"], check_only=True)` on this repository:

```
## isort
<frozen runpy>:88: UserWarning: Unable to parse file src\mcp_tools_py\code_checker_pytest\reporting.py due to 'charmap' codec can't encode character '✅' in position 23: character maps to <undefined>
... 11 files total
```

Every file containing `→`, `✅`, `≤`, `✨`, box-drawing characters or emoji is affected — 11 files here, 124 on the project where this was first found. isort exits 0 in every case, so the tool reports success.

## The masking is total, not partial

A probe file with **deliberately unsorted imports** and a `→` in its docstring, run through check mode:

```
rc=0   unparsable=1   errors=0
```

isort emitted **no** `Imports are incorrectly sorted` line. Check mode does not merely under-report a violation — it reports the opposite of one.

## Mode asymmetry (correction to the original report)

The same probe file, run through **apply** mode:

```
## isort
Fixing ...\tmp_isort_probe\with_arrow.py     <- sorted correctly
```

Apply mode processes these files fine. It prints no warnings because it never hits the error — not because it suppresses them.

Consequences:

- The fault's scope is `check_only=True` only.
- The original suggestion to "echo the warnings in apply mode" is not actionable: there is nothing to echo.
- **To verify before closing:** the originating CI incident on `mcp_coder` was attributed to apply mode leaving files unsorted. If apply mode works as shown here, that story needs another explanation — most likely a check-mode verification step in that flow. Re-check on that project.

## Impact

CI runs `isort --check --profile=black --float-to-top src tests` on Linux (`.github/workflows/ci.yml:109`), where the `charmap` codec is never selected and all files are parsed. The local check structurally cannot see what CI checks, while producing output identical to a passing run.

This repository is exposed today: 11 files are invisible to every local `run_format_code(check_only=True)`. It is currently benign — apply mode keeps them sorted, so CI passes — but the local check is providing false confidence, which is worse than providing none.

## Not fixable by invocation, environment, or flags

Original report (124-file project), both entry points:

| Invocation | exit | skipped |
|---|---|---|
| `.venv/Scripts/isort.exe --check-only src tests` | 0 | 124 |
| `.venv/Scripts/python.exe -m isort --check-only src tests` | 0 | 124 |

Environment variants, single-file probe, `python -m isort --check-only`:

| Variant | exit | unparsable |
|---|---|---|
| baseline (`PYTHONIOENCODING=utf-8`) | 0 | 1 |
| `PYTHONUTF8=1` | 0 | 1 |
| `PYTHONIOENCODING` cleared | 0 | 1 |

isort flag variants:

| Flags | exit | unparsable |
|---|---|---|
| `--check-only` | 0 | 1 |
| `--check-only --quiet` | 0 | 1 |
| `--check-only --diff` | 0 | 1 |
| `--check-only --verbose` | 0 | 1 |

Note `PYTHONIOENCODING=utf-8` is **already** set for every Python subprocess by `get_python_isolation_env()` in `mcp-coder-utils` — the existing mitigation is in place and insufficient. There is no environment-level or flag-level workaround for this server to apply.

## Implementation approach

Report the incomplete run truthfully. Four changes, in data-flow order.

**1. `src/mcp_tools_py/formatter/isort_runner.py`** — add `_parse_isort_unparsable_files(output)`, returning the paths from `Unable to parse file <path> due to <reason>`.

- Do a **per-line substring search**: locate `"Unable to parse file "` anywhere in the line, then slice to `rfind(" due to ")`. Do **not** split on whitespace — the paths are Windows paths and may contain spaces.
- Call it with the **untruncated `output` local**, mirroring `files_changed=_parse_isort_changed_files(output)` at `isort_runner.py:81`.
- The new function needs a Google-style docstring with a `Returns:` section — ruff `D` + `DOC` (google, preview) is enforced on `src/` (`pyproject.toml:86-101`; tests are exempt at :97).
- Set `success = result.return_code == 0 and not unparsable_files`.

**2. `src/mcp_tools_py/formatter/models.py`** — `FormatterResult` gains:

```python
unparsable_files: list[str] = dataclasses.field(default_factory=list)
```

`models.py` uses `import dataclasses` / `@dataclasses.dataclass` and never imports `field`, so a bare `field(...)` would be a `NameError`. The field goes **last** — it is the only defaulted field, so there is no ordering conflict — and must be added to the `Attributes:` block at `models.py:10-14`.

**3. Update the now-false docstrings** alongside the semantics change — `models.py:12` (`success: True when return_code == 0.`) and the `Returns:` block at `isort_runner.py:61-62`.

**4. `src/mcp_tools_py/formatter/formatter_tools.py`** — in `_format_results`, when the list is non-empty, emit a block **below** the `## isort` header and above isort's raw output:

```
## isort
ERROR: isort could not read 11 file(s) - they were NOT checked.
A clean result here does NOT mean CI will pass.
Known limitation (Windows, piped stdout).
  src\mcp_tools_py\code_checker_pytest\reporting.py
  tests\test_black_runner.py
  ... and 9 more
<isort's own output follows>
```

The block is emitted **generically** for any step whose `unparsable_files` is non-empty, interpolating the step name — no `step == "isort"` special case. Placing it below the header preserves the existing assertion `tests/test_formatter_tools.py:306` (`result.startswith("## isort")`).

`_format_results` already receives whole `FormatterResult` objects, so **no signature or type change is needed anywhere**; this is a one-line insert at `formatter_tools.py:110`.

`runner.py:63` fail-fast is **not** modified.

## Tests

Unit tests with mocked `execute_command`, matching the existing style in `tests/test_isort_runner.py`:

- parser extracts paths from a **verbatim** real warning line (prefix included)
- `success` is `False` when unparsable files are present despite `return_code == 0`
- `_format_results` renders the block, caps at 10, and reports the remainder

No real-isort integration test: the fault reproduces only on Windows with piped stdout, so such a test would pass vacuously on the Linux CI runners — its own instance of this bug.

## Constraints & Rationale

- **Match with a substring search, not `startswith`.** The real line is prefixed `<frozen runpy>:88: UserWarning: `. Both existing parsers in `isort_runner.py` (lines 39 and 41) use `startswith` and would match nothing.
- **Warnings arrive on stderr**, merged into `output` at `isort_runner.py:69-76` via `"\n".join`. A `warnings.warn` emission prints a *second*, indented source line after the warning; the parser must skip non-matching lines. A per-line substring search does this naturally, and the test fixture should include that trailing line to prove it is ignored.
- **Pin the test fixture to the single-line warning form.** The appendix below shows the message wrapped across two lines while *Observed behaviour* shows it on one. A genuinely wrapped message would break `rfind(" due to ")`.
- **The test fixture must not contain a live trigger character.** `tests/test_isort_runner.py` has none today; embedding a real `→`/`✅` would make that test file the 12th masked file — invisible to the very check it tests. (`tests/test_black_runner.py:25,118` already contains `✨ 🍰`, which is why it is one of the 11.)
- **Build the summary block in `_format_results`, not in the runner's `output`.** `_MAX_LINES = 200` truncates the runner's output; on a 124-file repo a summary appended there is pushed off the end by the very warnings it summarises. Parsing must stay before `_truncate_output` (it already is).
- **`FormatterResult` is a bare dataclass** with no defaults. The new field needs `dataclasses.field(default_factory=list)` or all four constructors break: `isort_runner.py:78-82`, `black_runner.py:77-81`, `tests/test_formatter_runner.py:18`, and `tests/test_formatter_tools.py:44`.
- **`_make_formatter_result()` (`tests/test_formatter_tools.py:43-45`) needs a new `unparsable_files` parameter**, not just a dataclass default, or the `_format_results` rendering test cannot be written.
- **`FormatterResult` is public API** — exported at `formatter/__init__.py:5` and listed in `__all__` at :7. Appending a defaulted field is backward-compatible. Nothing serializes, copies or compares it — no `dataclasses.asdict`/`replace`/`astuple`/`json.dumps`, and no test compares whole objects (every assertion is per-field) — so the dataclass `__eq__` change is inert.
- **`black_runner.py` duplicates rather than shares.** `_MAX_LINES` and `_truncate_output` are copy-pasted (`black_runner.py:9-24` is identical to `isort_runner.py:9-24`). Black needs no change beyond the dataclass default — but equally, nothing in black inherits this fix. This confirms the "black is out of scope" call.
- **`run_format_code` returns a plain `str`** (`formatter_tools.py:32-37`, registered via `server.py:84`). There is no structured status channel to the caller, so the user-visible verdict can only live in the text.
- **Do not re-litigate the environment workarounds** — the tables above are measured, not assumed.

## Decisions

| Topic | Decision | Rationale |
|---|---|---|
| Scope of the fault | Check mode only; apply mode works | Measured: apply mode correctly sorted a file containing `→` |
| CI story | Reframed; re-verify on `mcp_coder` before closing | Apply mode works here, so the original attribution cannot be right as stated |
| `success` semantics | `success = exit 0 AND no unparsable files` | `success` means "the step did its job over all target files". Leaving it `True` on a skipped-file run reproduces this very bug inside the data model, where future programmatic consumers inherit it |
| Fail-fast (`runner.py:63`) | Left untouched | Cannot trigger in check mode. If it ever triggered in apply mode, halting before black and printing "Formatting stopped due to errors in isort step" is correct behaviour. No new condition, no dead branch |
| "X of Y files processed" line | Dropped | isort reports no processed-file total in any mode (`--verbose` included). Its `Skipped N files` line counts config-based skips — a different thing. Counting `*.py` ourselves ignores `skip`, `extend_skip`, `src_paths` and `.gitignore`; a wrong ratio is worse than none |
| File list length | Cap at 10 + `... and N more` | isort's own warnings remain below the block as the full list |
| Mode gating | Applies in both modes | Cheap, and future-proof if isort ever starts warning in apply mode |
| Test strategy | Unit tests only, verbatim warning line as fixture | A real-isort test passes vacuously on Linux CI, where the fault cannot reproduce |
| Upstream isort bug | Reproducer captured in the appendix below | Filing/updating upstream is separate work |
| Guard against trigger characters in this repo | Not now | Speculative until the reporting fix proves insufficient; separate issue if wanted |
| Block wording scope | Emitted generically per step, name interpolated; no `step == "isort"` special case | `_format_results` loops uniformly at `formatter_tools.py:108-112`; a special case would be a dead branch for black, which the fail-fast decision already rejects |
| Docstring updates | `models.py:12`, the `Attributes:` block at `models.py:10-14`, and `isort_runner.py:61-62` updated with the semantics change | The `success` docstring becomes false the moment the semantics change; ruff `D`+`DOC` enforces the new parser's docstring anyway |
| Path delimiting | Substring `find`/`rfind` on `"Unable to parse file "` … `" due to "` | Windows paths may contain spaces, so whitespace splitting is unsafe |
| Step sizing | One commit | ~40 lines, no signature changes, no new module, no import-contract impact. Splitting leaves an intermediate state where isort reports `success=False` with no explanatory text — worse than either endpoint |

## Verification against the code

The plan was checked against the source across two analysis rounds. Confirmed accurate: the `startswith` claim (`isort_runner.py:39,41`), the bare-dataclass claim (`models.py:6-18`), the `_format_results` location (`formatter_tools.py:95-117`), the parse-before-truncate ordering, the 11-file count, the CI reference (`.github/workflows/ci.yml:109`), and the plain-`str` return of `run_format_code`.

Blast radius is complete: `.success` has exactly two production readers (`runner.py:63` and `formatter_tools.py:111`), and `FormatterResult` has exactly four constructors (listed above). No fifth exists.

Local isort and CI do not diverge on config — `pyproject.toml:79-81` supplies the same `--profile=black --float-to-top` settings that `ci.yml:109` passes explicitly.

## Out of scope

- **black** — exits 123 on files it cannot parse; already loud.
- **The isort bug itself** — not fixable here; see appendix.
- **A guard against trigger characters in this repo's sources** — separate issue if wanted.
- **Stale formatter documentation** — pre-existing and unrelated to this change: `docs/architecture/architecture.md:111-165` omits the `formatter` package and still says "8 tools total", `architecture.md:58` points at a non-existent `tools/format_all`, and `README.md:413-434` lists only pylint/pytest/mypy. Separate issue if wanted.

## Appendix — reproducer for the upstream isort report

Minimal case. File `probe.py`:

```python
"""Probe file: arrow → in docstring."""

import sys
import os
```

Run `python -m isort --check-only probe.py` on Windows with stdout piped:

```
<frozen runpy>:88: UserWarning: Unable to parse file probe.py due to 'charmap' codec
can't encode character '→' in position 21: character maps to <undefined>
```

Exit code 0. The unsorted imports are never reported. The same command in an interactive console processes the file correctly, as does `python -m isort probe.py` (apply mode) in the piped context.

Position 21 is the arrow's index in the file's first line, so the fault is in isort encoding *file content* to the console codec during the check path. Neither `PYTHONIOENCODING=utf-8` nor `PYTHONUTF8=1` changes it, which rules out `sys.stdout.encoding` and `locale.getpreferredencoding(False)` as the source.

Hypothesis (unverified — isort 9.0.1 ships as a compiled mypyc wheel with no readable source): `UnicodeEncodeError` subclasses `ValueError`, so isort's `except (OSError, ValueError): warn(f"Unable to parse file {file} due to {error}")` misfiles an output-encoding fault as a parse failure, and the handler's return value is not counted toward the exit code.

## Environment

- Windows 10/11, Python 3.11, isort 9.0.1 (compiled), black 26.5.1
- Reproduced on `mcp-tools-py` itself (11 files) and `mcp_coder` (124 files)

## Related

- #219 — mypy's 120s subprocess cap
- #220 — the tool-availability probe's 10s timeout

All three share a shape: a checker fails for environmental reasons and the failure is reported in a way that points away from the real cause.
